[TRTLLM-6756][feat] Update BeamSearch for TorchSampler

stnie opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Enhanced beam search sampling with improved token padding and metadata handling.
  * Added support for tensor-based temperature parameters in beam search operations.

* **Bug Fixes**
  * Improved beam history and token tracking for beam search decoding.
  * Fixed logprob extraction and token padding for finished beams.

* **Tests**
  * Updated beam search tests to validate new padding semantics and CUDA synchronization behavior.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

Follow up on PR #8509

- Enable BeamSearch when using FlashInferGroupedStrategySampler
- Update handling of early finished beams
- simplified finalize_beams
- other smaller refactoring changes
- update test_beam_search accordingly

## Test Coverage

Updated test_beam_search.py to conform with the introduced changes

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

stnie committed:
[TRTLLM-6756][feat]  Update BeamSearch for TorchSampler

- Enable BeamSearch when using FlashInferGroupedStrategySampler
- Update handling of early finished beams
- simplified finalize_beams
- other smaller refactoring changes
- update test_beam_search accordingly

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampling_utils_flashinfer.py:
```diff
@@ -551,6 +622,111 @@ def sample(
+    return BeamSearchSampleOnly_local
+
+
+def create_beam_search_with_probs_cls(
```
- I suggest to deduplicate this by adding kwarg `with_probs: bool` and then `class BeamSearchSpecialized(_StrategyImpls.BeamSearchWithProbs if with_probs else _StrategyImpls.BeamSearchSampleOnly): ...`
- nit: use Pascal case for class name
- nit: prefix module-level implementation details with underscore, e.g.,  `_create_beam_search_sample_only_cls`

ixlmar commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -1466,86 +1478,70 @@ def _finalize_beam(
-                    )
-                case _:
-                    metadata = None
+            metadata_type = get_metadata_type_for_group_fn(key.strategy)
```
Rename `key.strategy -> key.strategy_key` (it is not always a `Strategy`)

ixlmar commented on tests/unittest/_torch/sampler/test_beam_search.py:
The tests should also parametrize `disable_flashinfer_sampling`.

stnie commented on tensorrt_llm/_torch/pyexecutor/sampling_utils_flashinfer.py:
```diff
@@ -346,6 +355,63 @@ def sample(
+            assert group_metadata is not None and isinstance(group_metadata, BeamSearchMetadata), (
+                "BeamSearchMetadata is required for beam_search_sampling_batch"
+            )
+            assert all(self._beam_width_in == self._beam_width_in[0]), (
```
I adjusted this

stnie committed:
[TRTLLM-6756][chore] Fix smaller issues within the beam search update

- Unified implementation of BeamSearchWithProbs and BeamSearchSampleOnly into BeamSearchMixin
- Simplified some checks accordingly
- moved temperature scaling and logit selection in BeamSearchMixin to  prepare_logits_with_temperature.
- removed duplicated test in test_beam_search.py
- extend test_beam_search to also parametrize disable_flashinfer_sampling

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #27139](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/27139/) [ run ] completed with state `SUCCESS`. Commit: `f24b35c` 
[/LLM/main/L0_MergeRequest_PR pipeline #20708](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/20708/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/20708/test-results/rerun_report.html) for details.
